File: src/main/services/terminal/strategies/cmd.ts
import { spawn } from 'child_process'
import type { TerminalStrategy, EnvObject } from './base'
import { escapeEnvValue, sanitizePath } from './escape'

export class CmdStrategy implements TerminalStrategy {
  name = 'Command Prompt'

  async isAvailable(): Promise<boolean> {
    return process.platform === 'win32'
  }

  async launch(path: string, env: EnvObject, cliCommand?: string): Promise<void> {
    const safePath = sanitizePath(path)
    const envSetCommands = Object.entries(env)
      .map(([key, value]) => `set "${key}=${escapeEnvValue(value)}"`)
      .join(' && ')

    // Use pushd to support UNC paths (e.g. \\psf\Home\...)
    let finalCommand = `pushd "${safePath}"`
    if (envSetCommands) {
      finalCommand += ` && ${envSetCommands}`
    }
    if (cliCommand) {
      finalCommand += ` && ${cliCommand}`
    }

    spawn('cmd.exe', ['/k', finalCommand], {
      detached: true,
      stdio: 'ignore',
    }).unref()
  }
}

File: src-tauri/src/terminal/cmd.rs
use super::{EnvObject, TerminalStrategy};
use std::process::Command;

pub struct CmdStrategy;

impl TerminalStrategy for CmdStrategy {
    fn name(&self) -> &str { "CMD" }

    fn is_available(&self) -> bool {
        cfg!(target_os = "windows")
    }

    fn launch(&self, path: &str, env: &EnvObject, cli_command: &str) -> Result<(), String> {
        let env_sets: String = env.iter()
            .map(|(k, v)| format!("set {}={}", k, v))
            .collect::<Vec<_>>()
            .join(" && ");

        let full_command = if env_sets.is_empty() {
            format!("cd /d \"{}\" && {}", path, cli_command)
        } else {
            format!("cd /d \"{}\" && {} && {}", path, env_sets, cli_command)
        };

        Command::new("cmd")
            .args(["/C", "start", "cmd", "/K", &full_command])
            .spawn()
            .map_err(|e| format!("Failed to launch CMD: {}", e))?;

        Ok(())
    }
}
